From a49e5fdcad2917b64ac4d32c999ce1ab2d084c93 Mon Sep 17 00:00:00 2001
From: Xi Ruoyao <xry111@xry111.site>
Date: Sat, 21 Jun 2025 17:59:23 +0800
Subject: [PATCH 10/63] AOSCOS: loonggpu: Adapt for drm_sched_init struct
 params

Since kernel commit 796a9f55a8d1 ("drm/sched: Use struct for
drm_sched_init() params"), most parameters of drm_sched_init() are
packed into a struct.

Signed-off-by: Xi Ruoyao <xry111@xry111.site>
---
 conftest.sh               | 18 ++++++++++++++++++
 include/loonggpu_helper.h | 22 +++++++++++++++++-----
 loonggpu/loonggpu.Kbuild  |  1 +
 3 files changed, 36 insertions(+), 5 deletions(-)

diff --git a/conftest.sh b/conftest.sh
index 3f7b54d..18b64d1 100644
--- a/conftest.sh
+++ b/conftest.sh
@@ -1217,6 +1217,24 @@ compile_test() {
             compile_check_conftest "$CODE" "LG_DRM_SCHED_INIT_HAS_SUBMIT_WQ" "" "types"
         ;;
 
+        drm_sched_init_has_struct_arg)
+            #
+            # Determine if drm_sched_init uses struct drm_sched_init_args arg.
+            #
+            # Changed by commit 796a9f55a8d1d85387b973df9a06cbf4bc2d6327
+            # drm/sched: Use struct for drm_sched_init() params
+            #
+            CODE="
+            #include <drm/gpu_scheduler.h>
+            typeof(drm_sched_init) conftest_drm_sched_init_has_submit_wq;
+            int conftest_drm_sched_init_has_submit_wq(struct drm_gpu_scheduler *sched,
+                                                      const struct drm_sched_init_args *args
+                                                      ) {
+                return 0;
+            }"
+            compile_check_conftest "$CODE" "LG_DRM_SCHED_INIT_HAS_STRUCT_ARG" "" "types"
+        ;;
+
         drm_sched_job_init_has_credits)
             #
             # Determine if drm_sched_job_init has credits arg.
diff --git a/include/loonggpu_helper.h b/include/loonggpu_helper.h
index c40d881..91ff130 100644
--- a/include/loonggpu_helper.h
+++ b/include/loonggpu_helper.h
@@ -738,7 +738,7 @@ static inline long lg_get_user_pages(uint64_t userptr, unsigned num_pages,
 static inline struct drm_sched_rq *lg_sched_to_sched_rq(struct drm_gpu_scheduler *sched,
 					enum drm_sched_priority priority)
 {
-#if defined(LG_DRM_SCHED_INIT_HAS_DEVICE_RQ) || defined (LG_DRM_SCHED_INIT_HAS_SUBMIT_WQ)
+#if defined(LG_DRM_SCHED_INIT_HAS_DEVICE_RQ) || defined (LG_DRM_SCHED_INIT_HAS_SUBMIT_WQ) || defined (LG_DRM_SCHED_INIT_HAS_STRUCT_ARG)
 	return sched->sched_rq[priority];
 #else
 	return &sched->sched_rq[priority];
@@ -751,7 +751,19 @@ static inline int lg_drm_sched_init(struct loonggpu_ring *ring,
 				unsigned hang_limit,
 				long timeout, struct loonggpu_device *adev)
 {
-#if defined(LG_DRM_SCHED_INIT_HAS_DEVICE)
+#if defined (LG_DRM_SCHED_INIT_HAS_STRUCT_ARG)
+	const struct drm_sched_init_args args = {
+		.ops = ops,
+		.num_rqs = DRM_SCHED_PRIORITY_COUNT,
+		.credit_limit = num_hw_submission,
+		.hang_limit = hang_limit,
+		.timeout = timeout,
+		.name = ring->name,
+		.dev = adev->dev,
+	};
+
+	return drm_sched_init(&ring->sched, &args);
+#elif defined(LG_DRM_SCHED_INIT_HAS_DEVICE)
 	return drm_sched_init(&ring->sched, ops,
 			   num_hw_submission, hang_limit,
 			   timeout, NULL,
@@ -1172,7 +1184,7 @@ static inline bool lg_ring_sched_thread_avai(struct loonggpu_ring *ring)
 {
 	if (!ring)
 		return false;
-#if defined(LG_DRM_SCHED_INIT_HAS_SUBMIT_WQ)
+#if defined(LG_DRM_SCHED_INIT_HAS_SUBMIT_WQ) || defined(LG_DRM_SCHED_INIT_HAS_STRUCT_ARG) 
 	if (!drm_sched_wqueue_ready(&ring->sched))
 		return false;
 #else
@@ -1184,7 +1196,7 @@ static inline bool lg_ring_sched_thread_avai(struct loonggpu_ring *ring)
 
 static inline void lg_ring_sched_thread_park(struct loonggpu_ring *ring)
 {
-#if defined(LG_DRM_SCHED_INIT_HAS_SUBMIT_WQ)
+#if defined(LG_DRM_SCHED_INIT_HAS_SUBMIT_WQ) || defined(LG_DRM_SCHED_INIT_HAS_STRUCT_ARG)
 	drm_sched_wqueue_stop(&ring->sched);
 #else
 	kthread_park(ring->sched.thread);
@@ -1193,7 +1205,7 @@ static inline void lg_ring_sched_thread_park(struct loonggpu_ring *ring)
 
 static inline void lg_ring_sched_thread_unpark(struct loonggpu_ring *ring)
 {
-#if defined(LG_DRM_SCHED_INIT_HAS_SUBMIT_WQ)
+#if defined(LG_DRM_SCHED_INIT_HAS_SUBMIT_WQ) || defined(LG_DRM_SCHED_INIT_HAS_STRUCT_ARG)
 	drm_sched_wqueue_start(&ring->sched);
 #else
 	kthread_unpark(ring->sched.thread);
diff --git a/loonggpu/loonggpu.Kbuild b/loonggpu/loonggpu.Kbuild
index a2e65cf..4fa5b00 100644
--- a/loonggpu/loonggpu.Kbuild
+++ b/loonggpu/loonggpu.Kbuild
@@ -196,3 +196,4 @@ LG_CONFTEST_TYPE_COMPILE_TESTS += dma_resv_replace_fences
 LG_CONFTEST_TYPE_COMPILE_TESTS += mode_config_has_fb_modifiers
 LG_CONFTEST_TYPE_COMPILE_TESTS += mmu_notifier_ops_has_free_notifier
 LG_CONFTEST_TYPE_COMPILE_TESTS += pwm_add_table
+LG_CONFTEST_TYPE_COMPILE_TESTS += drm_sched_init_has_struct_arg
-- 
2.52.0

